File: Sources/KairosRPC/Envelope.swift
import Foundation

/// A request line: `{"method":"...","params":{...}}`.
public struct RequestEnvelope: Codable, Sendable, Equatable {
    public let method: Method
    public let params: JSONValue

    public init(method: Method, params: JSONValue) {
        self.method = method
        self.params = params
    }
}

/// A response line: `{"result":...}` or `{"error":{"code","message"}}`.
public enum ResponseEnvelope: Codable, Sendable, Equatable {
    case result(JSONValue)
    case error(RPCError)

    private enum CodingKeys: String, CodingKey { case result, error }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        if let err = try? c.decode(RPCError.self, forKey: .error) {
            self = .error(err)
        } else if let res = try? c.decode(JSONValue.self, forKey: .result) {
            self = .result(res)
        } else {
            throw RPCDecodeError.malformedResponse
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case .result(let v): try c.encode(v, forKey: .result)
        case .error(let e): try c.encode(e, forKey: .error)
        }
    }
}

File: libs/codec/src/envelope.rs
//! Request/response envelopes. The wire shape is `{"method","params"}` for
//! requests and exactly one of `{"result":...}` / `{"error":{"code","message"}}`
//! for responses — mirrors Swift `RequestEnvelope` / `ResponseEnvelope`.

use serde::ser::SerializeMap;
use serde::{Deserialize, Serialize};

use crate::{Method, RpcError, Value};

/// `{"method":"...","params":{...}}`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct RequestEnvelope {
    pub method: Method,
    pub params: Value,
}

impl RequestEnvelope {
    pub fn new(method: Method, params: Value) -> Self {
        Self { method, params }
    }
}

/// `{"result":...}` or `{"error":{...}}` — exactly one key present.
#[derive(Debug, Clone, PartialEq)]
pub enum ResponseEnvelope {
    Result(Value),
    Error(RpcError),
}

impl Serialize for ResponseEnvelope {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut map = serializer.serialize_map(Some(1))?;
        match self {
            ResponseEnvelope::Result(v) => map.serialize_entry("result", v)?,
            ResponseEnvelope::Error(e) => map.serialize_entry("error", e)?,
        }
        map.end()
    }
}

impl<'de> Deserialize<'de> for ResponseEnvelope {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let map = serde_json::Map::deserialize(deserializer)?;
        if let Some(err) = map.get("error") {
            let error =
                serde_json::from_value::<RpcError>(err.clone()).map_err(serde::de::Error::custom)?;
            return Ok(ResponseEnvelope::Error(error));
        }
        if let Some(result) = map.get("result") {
            return Ok(ResponseEnvelope::Result(result.clone()));
        }
        Err(serde::de::Error::custom(
            "response has neither result nor error",
        ))
    }
}
